### PR TITLE
Implement routing

### PR DESCRIPTION
**Pull request by janhohenheim**

### CODE REVIEW

File: Routing/Parsing/PathParser.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    internal class PathParser : IPathParser
+    {
+        /// <summary>
+        /// ABNF spec from <a href="https://tools.ietf.org/html/rfc3986">RFC3986</a>:
+        /// <code>
+        /// segment = *pchar
+        /// pchar = unreserved / pct-encoded / sub-delims / ":" / "@"
+        /// unreserved  = ALPHA / DIGIT / "-" / "." / "_" / "~"
+        /// pct-encoded = "%" HEXDIG HEXDIG
+        /// sub-delims  = "!" / "$" / "&" / "'" / "(" / ")" / "*" / "+" / "," / ";" / "="
```
Comment (tautropfli) [inline Routing/Parsing/PathParser.cs]:
```suggestion
        /// sub-delims  = "!" / "$" / "&amp;" / "'" / "(" / ")" / "*" / "+" / "," / ";" / "="
```

File: Routing.Test/Routing.Test.csproj [hunk 2]
```diff
@@ -2,6 +2,11 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
     <TargetFramework>netcoreapp3.0</TargetFramework>
+    <LangVersion>8.0</LangVersion>
+    <Nullable>enable</Nullable>
+    <NullableContextOptions>enable</NullableContextOptions>
```
Comment (tautropfli) [inline Routing.Test/Routing.Test.csproj]: Remove this as it is no longer needed.

File: Routing.Test/RoutingTest.cs [hunk 3]
```diff
@@ -256,6 +257,46 @@ public void ParsesMultipleRouteParams()
...
+            }, CreateExpectedParams(("bar", "bar"), ("baz", "baz")), expectedRoute);
+        }
+
+        [Fact]
+        public static void FirstRegisteredAmbiguousRouteIsPreferred()
+        {
+            const string name = "foo";
+            var ambiguousRoute = $"{RegisteredRoute}/{{bar}}";
+
+            AssertRouteWasCalledWithParams(routeRegistry =>
```
Comment (tautropfli) [inline Routing.Test/RoutingTest.cs]: Put the non-function parameters first.

File: Routing/Parsing/SegmentParser.cs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        public IEnumerable<ISegmentVariant>? Parse(string route)
+        {
+            var segments = ParseSegments(route);
+
+            if (segments is null || segments.Any(segment => segment is null))
+            {
+                return null;
+            }
+
+            CheckForDuplicatedParameters(segments!);
```
Comment (tautropfli) [inline Routing/Parsing/SegmentParser.cs:30]: Is the `!` really needed here?

File: Routing/RouteRegistry.cs [hunk 5]
```diff
@@ -1,37 +1,187 @@
...
+            {
+                [key] = value
+            };
         }
+
+
         public IRouteRegistry<TResponse, TRequest> Register(HttpMethod method, string route, HandleRequest<TResponse, TRequest> handleRequest)
         {
-            _registeredRoutes[route] = handleRequest;
+            var segments = ParseRoute(route);
```
Comment (tautropfli) [inline Routing/RouteRegistry.cs]: Should we extract the data structure and the insert / remove operations into a separate class?

File: Routing/RouteRegistry.cs [hunk 6]
```diff
@@ -1,37 +1,187 @@
...
+
             return this;
         }
 
+        private static SegmentNode<TResponse, TRequest> FindOrInsertNode(ICollection<SegmentNode<TResponse, TRequest>> collection, ISegmentVariant segment)
+        {
+            var existingNode = collection.FirstOrDefault(element => element.Matcher.Equals(segment));
+            return existingNode ?? InsertNode(collection, segment);
+        }
+        private static SegmentNode<TResponse, TRequest> InsertNode(ICollection<SegmentNode<TResponse, TRequest>> collection, ISegmentVariant segment)
```
Comment (tautropfli) [inline Routing/RouteRegistry.cs]: Missing newline

File: Routing/Routing.csproj [hunk 7]
```diff
@@ -1,7 +1,12 @@
-<?xml version="1.0" encoding="utf-8"?>
+﻿<?xml version="1.0" encoding="utf-8"?>
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFramework>netstandard2.0</TargetFramework>
+    <TargetFramework>netstandard2.1</TargetFramework>
+    <LangVersion>8.0</LangVersion>
+    <Nullable>enable</Nullable>
+    <NullableContextOptions>enable</NullableContextOptions>
```
Comment (tautropfli) [inline Routing/Routing.csproj]: Remove this as it is no longer needed.

File: paket.dependencies [hunk 8]
```diff
@@ -5,3 +5,5 @@ nuget Funcky ~> 1.2.0
 nuget Microsoft.NET.Test.Sdk ~> 16.2.0
 nuget xunit ~> 2.4.0
 nuget xunit.runner.visualstudio ~> 2.4.0
+nuget Equals.Fody ~> 1.9.5
+nuget Fody ~> 4.0
```
Comment (tautropfli) [inline paket.dependencies]: Use the latest version and make the dependencies private.
See:
https://github.com/messerli-informatik-ag/file-opening-builder/blob/c3abf3f156006e9737430192c66e5d6f4e346134/paket.dependencies#L5-L6

File: Routing/SegmentVariant/Root.cs [hunk 9]
```diff
@@ -0,0 +1,8 @@
+﻿namespace Routing.SegmentVariant
+{
+    [Equals]
+    internal class Root : ISegmentVariant
+    {
+
```
Comment (tautropfli) [inline Routing/SegmentVariant/Root.cs:9]: Remove this blank line

File: Routing/SegmentVariant/ISegmentVariant.cs [hunk 10]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Routing.SegmentVariant
+{
+    internal interface ISegmentVariant
+    {
+
```
Comment (tautropfli) [inline Routing/SegmentVariant/ISegmentVariant.cs]: Remove this blank line

**Review by tautropfli — CHANGES_REQUESTED** [review]

File: Routing/Parsing/SegmentParser.cs (hunk 4 above)
Comment (janhohenheim) [reply Routing/Parsing/SegmentParser.cs:30]: yup

**Review by dngroth — DISMISSED** [review]

File: Routing/RouteRegistry.cs [hunk 11]
```diff
@@ -1,37 +1,187 @@
...
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
 using System.Net.Http;
+using Routing.Parsing;
+using Routing.SegmentVariant;
 
 namespace Routing
 {
     public class RouteRegistry<TResponse, TRequest> : IRouteRegistry<TResponse, TRequest>
```
Comment (tautropfli) [inline Routing/RouteRegistry.cs]: `TRequest` and `TResponse` should be swapped.

**Review by tautropfli — CHANGES_REQUESTED** [review]

File: Routing.Test/Routing.Test.csproj (hunk 2 above)
Comment (janhohenheim) [reply Routing.Test/Routing.Test.csproj]: https://github.com/messerli-informatik-ag/routing/pull/6/commits/370cb929460c1039c268cf4b22e4776d4678ac27

File: Routing/Routing.csproj (hunk 7 above)
Comment (janhohenheim) [reply Routing/Routing.csproj]: https://github.com/messerli-informatik-ag/routing/pull/6/commits/370cb929460c1039c268cf4b22e4776d4678ac27

File: Routing/SegmentVariant/ISegmentVariant.cs (hunk 10 above)
Comment (janhohenheim) [reply Routing/SegmentVariant/ISegmentVariant.cs]: https://github.com/messerli-informatik-ag/routing/pull/6/commits/ede78780d40dc6aa81b6837a3536a9538b64f89d

File: Routing/SegmentVariant/Root.cs (hunk 9 above)
Comment (janhohenheim) [reply Routing/SegmentVariant/Root.cs:9]: https://github.com/messerli-informatik-ag/routing/pull/6/commits/ede78780d40dc6aa81b6837a3536a9538b64f89d

File: Routing/RouteRegistry.cs (hunk 6 above)
Comment (janhohenheim) [reply Routing/RouteRegistry.cs]: https://github.com/messerli-informatik-ag/routing/pull/6/commits/18b4339fc5341d72acd1f56078ba4c3d66b875b7

File: paket.dependencies (hunk 8 above)
Comment (janhohenheim) [reply paket.dependencies]: https://github.com/messerli-informatik-ag/routing/pull/6/commits/fed414c46866844f7309b2214beb1e5a38d461ed

File: Routing/RouteRegistry.cs (hunk 11 above)
Comment (janhohenheim) [reply Routing/RouteRegistry.cs]: https://github.com/messerli-informatik-ag/routing/pull/6/commits/2498e174f580c386d1be3efa81eb09a9db22a977

File: Routing.Test/RoutingTest.cs (hunk 3 above)
Comment (janhohenheim) [reply Routing.Test/RoutingTest.cs]: https://github.com/messerli-informatik-ag/routing/pull/6/commits/2498e174f580c386d1be3efa81eb09a9db22a977

File: Routing/RouteRegistry.cs (hunk 5 above)
Comment (janhohenheim) [reply Routing/RouteRegistry.cs]: https://github.com/messerli-informatik-ag/routing/pull/6/commits/9190e6d30a507ffc52847bdb0ff7d1075cab4468

File: Routing/SegmentRegistryFacadeImplementation/Router.cs [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+            return node.LiteralChildren
+                .Concat(node.ParameterChildren)
+                .Select(child => Match(child, method, tail, currentParameters))
+                .FirstOrDefault(child => child != null);
+        }
+        private static Func<TRequest, TResponse> CurryParameters(
+            HandleRequest<TRequest, TResponse> handleRequest,
+            IDictionary<string, string> parameters) =>
+            request => handleRequest(request, parameters);
+        private static bool NodeMatchesSegment(SegmentNode<TRequest, TResponse> node, string segment) =>
```
Comment (tautropfli) [inline Routing/SegmentRegistryFacadeImplementation/Router.cs:74]: Missing newline

File: Routing/SegmentVariant/Literal.cs [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+    internal class Literal : ISegmentVariant
+    {
+        internal Literal(string identifier)
+        {
+            Identifier = identifier;
+        }
+
+        internal string Identifier { get; }
+
+        public static bool operator==(Literal left, Literal right) => Operator.Weave(left, right);
```
Comment (tautropfli) [inline Routing/SegmentVariant/Literal.cs]: Missing spaces around the operator

**Review by tautropfli — CHANGES_REQUESTED** [review]

File: Routing/SegmentRegistryFacadeImplementation/Router.cs (hunk 12 above)
Comment (janhohenheim) [reply Routing/SegmentRegistryFacadeImplementation/Router.cs:74]: https://github.com/messerli-informatik-ag/routing/pull/6/commits/f877a13ceca04a503a59256b3e286f6a639b18c0

File: Routing/SegmentVariant/Literal.cs (hunk 13 above)
Comment (jnferner) [reply Routing/SegmentVariant/Literal.cs]: https://github.com/messerli-informatik-ag/routing/pull/6/commits/f877a13ceca04a503a59256b3e286f6a639b18c0

Comment (janhohenheim) [reply Routing/SegmentVariant/Literal.cs]: https://github.com/messerli-informatik-ag/routing/pull/6/commits/cb332a60a6bb41fd2d749a6ea5025562a38fcb7d

**Review by tautropfli — CHANGES_REQUESTED** [review]
Mark classes as `sealed`.

**Comment by janhohenheim** [comment]
@bash https://github.com/messerli-informatik-ag/routing/pull/6/commits/943c2f618fcd3d2e93f87636bd1d7a4af94b45c5

**Review by tautropfli — APPROVED** [review]

**Pull request merged**